### PR TITLE
promote `collect` macro as a map+filter replacement

### PR DESCRIPTION
**Pull request by narimiran**

### CODE REVIEW

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -361,7 +368,11 @@ proc map*[T, S](s: openArray[T], op: proc (x: T): S {.closure.}):
   ## Since the input is not modified you can use it to
   ## transform the type of the elements in the input container.
   ##
+  ## Instead of using `map` and `filter`, consider using the `collect` macro
+  ## from `sugar` module.
```
Comment (Araq) [inline lib/pure/collections/sequtils.nim]: "from THE `sugar` module"

Comment (narimiran) [inline lib/pure/collections/sequtils.nim]:
```suggestion
  ## from the `sugar` module.
```

**Comment by bluenote10** [comment]
This is not "functional style programming".

**Comment by Araq** [comment]
> This is not "functional style programming".

Yeah, it's better as it doesn't confuse (lazy) functions with control flow aspects.

**Comment by bluenote10** [comment]
Just pointing out that if the docs list this `collect` under "functional programming" you can easily disappoint functional programmers. It may be "better", but this ugly `untyped -> untyped` "function" doesn't get the idea of functional programming, exactly because it introduces imperative style control flow.

In functional programming `map` + `filter` is `flatMap`.

**Comment by Araq** [comment]
Well, there is nothing wrong with teaching functional programmers.

**Pull request merged**